### PR TITLE
No ACS URL Test

### PR DESCRIPTION
**Pull request by blen-desta**
- Removes `redirect-authn-request.xml` and uses `AuthnRequestBuilder` to build redirect messages.
- Adds a test for:

```
The identity provider MUST have some means to establish that this location is in 
fact controlled by the service provider. A service provider MAY indicate the 
SAML binding and the specific assertion consumer service to use in its 
<AuthnRequest> and the identity provider MUST honor them if it can.
```

We already have verifiers verifying the destination in the response is the provided ACS url (in the request). This test is meant to verify that the IdP can respond with the correct destination without the SP providing an ACS url.

### CODE REVIEW

File: test/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt [hunk 1]
```diff
@@ -42,42 +36,26 @@ class TestCommon {
...
-
-        private val DEPLOY_CL = getDeployDirClassloader()
+        var acsUrl : String? = ""
 
         /**
-         * Generates and returns a POST Authn Request
+         * Returns the Assertion Consumer Service URL for the provided binding
          */
-        fun generateAndRetrieveAuthnRequest(): String {
-            OpenSAMLUtil.initSamlEngine()
```
Comment (brjeter) [inline test/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt]: ❓ Are we sure we don't need this anymore?

Comment (blen-desta) [reply test/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt]: We moved the generation of the request to the tests themselves. We are already using an AuthnRequest builder so tests should have the "choice" on what fields to provide. I deleted the method but the logic is in the tests.

File: test/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt [hunk 2]
```diff
@@ -42,42 +36,26 @@ class TestCommon {
...
-                nameIDPolicy = NameIDPolicyBuilder().buildObject().apply {
-                    allowCreate = true
-                    format = SAML2Constants.NAMEID_FORMAT_PERSISTENT
-                    spNameQualifier = SP_ISSUER
-                }
-            }
-
-            SimpleSign().signSamlObject(authnRequest)
+        fun acsUrl(binding: SamlProtocol.Binding): String? {
+            val spInfo = spMetadata[SP_ISSUER]
```
[the remark this answers is not in the document]
Comment (blen-desta) [reply test/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt]: It depends on the binding and there are verifiers that check the constant without knowing what binding we're using

File: test/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt (hunk 1 above)
Comment (brjeter) [reply test/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt]: Sorry for not being more clear... I meant this `initSamlEngine` method call. I don't see it being moved to somewhere else. It's just removed.

Comment (blen-desta) [reply test/common/src/main/kotlin/org/codice/compliance/utils/TestCommon.kt]: Oh, I see. Everything seemed like it was working without it. I'm honestly not sure what that does

File: test/common/src/main/kotlin/org/codice/compliance/verification/binding/PostBindingVerifier.kt [hunk 3]
```diff
@@ -102,12 +103,12 @@ class PostBindingVerifier(private val response: IdpPostResponseDecorator) {
         val destination = response.responseDom.attributes.getNamedItem("Destination")?.nodeValue
         val signatures = response.responseDom.allChildren("Signature")
 
-        if (signatures.isNotEmpty() && destination != TestCommon.ACS_URL) {
+        if (signatures.isNotEmpty() && destination != acsUrl[SamlProtocol.Binding.HTTP_POST]) {
```
Comment (coyotesqrl) [inline test/common/src/main/kotlin/org/codice/compliance/verification/binding/PostBindingVerifier.kt]: ✏️ Totally up to you, but a static import of `HTTP_POST` would be less wordy.

File: test/common/src/main/kotlin/org/codice/compliance/verification/binding/PostBindingVerifier.kt [hunk 4]
```diff
@@ -102,12 +103,12 @@ class PostBindingVerifier(private val response: IdpPostResponseDecorator) {
...
             throw SAMLComplianceException.createWithPropertyNotEqualMessage(
                     SAMLBindings_3_5_5_2_a,
                     "Destination",
                     destination,
-                    TestCommon.ACS_URL)
+                    acsUrl[SamlProtocol.Binding.HTTP_POST])
```
Comment (coyotesqrl) [inline test/common/src/main/kotlin/org/codice/compliance/verification/binding/PostBindingVerifier.kt]: ✏️ Again, a static import of `HTTP_POST` would be less wordy.

File: test/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt [hunk 5]
```diff
@@ -203,12 +204,12 @@ class RedirectBindingVerifier(private val response: IdpRedirectResponseDecorator
         val destination = response.responseDom.attributes.getNamedItem("Destination")?.nodeValue
         val signatures = response.responseDom.allChildren("Signature")
 
-        if (signatures.isNotEmpty() && destination != TestCommon.ACS_URL) {
+        if (signatures.isNotEmpty() && destination != acsUrl[SamlProtocol.Binding.HTTP_REDIRECT]) {
```
Comment (coyotesqrl) [inline test/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt]: ✏️ Similarly, a static import of `HTTP_REDIRECT ` would be less wordy.

File: test/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt [hunk 6]
```diff
@@ -203,12 +204,12 @@ class RedirectBindingVerifier(private val response: IdpRedirectResponseDecorator
...
             throw SAMLComplianceException.createWithPropertyNotEqualMessage(
                     SAMLBindings_3_5_5_2_a,
                     "Destination",
                     destination,
-                    TestCommon.ACS_URL)
+                    acsUrl[SamlProtocol.Binding.HTTP_REDIRECT])
```
Comment (coyotesqrl) [inline test/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt]: ✏️ Again, a static import of `HTTP_REDIRECT ` would be less wordy.

File: test/common/src/main/kotlin/org/codice/compliance/verification/core/ResponseProtocolVerifier.kt [hunk 7]
```diff
@@ -84,11 +83,11 @@ class ResponseProtocolVerifier(val response: Node, val id: String) {
         verifyTimeValues(response.attributes.getNamedItem("IssueInstant"), SAMLCore_3_2_2_d)
 
         val destination = response.attributes?.getNamedItem("Destination")?.textContent
-        if (destination != ACS_URL)
+        if (destination != acsUrl)
             throw SAMLComplianceException.createWithPropertyNotEqualMessage(SAMLCore_3_2_2_e,
                     "Destination",
                     destination,
-                    ACS_URL)
+                    acsUrl ?: "NO_ACS")
```
Comment (coyotesqrl) [inline test/common/src/main/kotlin/org/codice/compliance/verification/core/ResponseProtocolVerifier.kt]: ✏️ When I made my quick patch this morning, I just shoved this in so we'd have something to display. We might want to make it a constant or use some other string.

File: test/idp/src/main/kotlin/org/codice/compliance/web/sso/PostLoginTest.kt [hunk 8]
```diff
@@ -56,21 +79,34 @@ class PostLoginTest : StringSpec() {
...
             idpResponse.bindingVerifier().verify()
 
             val responseDom = idpResponse.responseDom
-
             CoreVerifier(responseDom).verify()
-
-            ResponseProtocolVerifier(responseDom, TestCommon.ID).verify()
-
-            SingleSignOnProfileVerifier(responseDom).verify()
+            ResponseProtocolVerifier(responseDom, TestCommon.ID, acsUrl[SamlProtocol.Binding.HTTP_POST]).verify()
```
Comment (coyotesqrl) [inline test/idp/src/main/kotlin/org/codice/compliance/web/sso/PostLoginTest.kt]: ✏️ More places where static imports would shorten.

File: test/idp/src/main/kotlin/org/codice/compliance/web/sso/PostLoginTest.kt [hunk 9]
```diff
@@ -56,21 +79,34 @@ class PostLoginTest : StringSpec() {
...
+            SingleSignOnProfileVerifier(responseDom, acsUrl[SamlProtocol.Binding.HTTP_POST]).verify()
         }
 
         "POST AuthnRequest With Relay State Test" {
-            val authnRequest = generateAndRetrieveAuthnRequest()
-            val encodedRequest = Encoder.encodePostMessage(authnRequest, EXAMPLE_RELAY_STATE)
+            val authnRequest = AuthnRequestBuilder().buildObject().apply {
```
Comment (coyotesqrl) [inline test/idp/src/main/kotlin/org/codice/compliance/web/sso/PostLoginTest.kt]: ✏️ This block and the ones on line 52 and 135 are either identical or nearly so?

Can we extract to a setup method or extract a function to do the creation?

File: test/idp/src/main/kotlin/org/codice/compliance/web/sso/PostLoginTest.kt [hunk 10]
```diff
@@ -91,10 +127,50 @@ class PostLoginTest : StringSpec() {
...
+                protocolBinding = SamlProtocol.POST_BINDING
+                nameIDPolicy = NameIDPolicyBuilder().buildObject().apply {
+                    allowCreate = true
+                    format = SAML2Constants.NAMEID_FORMAT_PERSISTENT
+                    spNameQualifier = TestCommon.SP_ISSUER
+                }
+            }.apply { SimpleSign().signSamlObject(this) }
+
+            //set the acs url for the verifier to verify against
+            acsUrl[SamlProtocol.Binding.HTTP_POST]
```
Comment (coyotesqrl) [inline test/idp/src/main/kotlin/org/codice/compliance/web/sso/PostLoginTest.kt]: ✏️ This line does nothing.

File: test/idp/src/main/kotlin/org/codice/compliance/web/sso/PostLoginTest.kt [hunk 11]
```diff
@@ -91,10 +127,50 @@ class PostLoginTest : StringSpec() {
...
+
+            response.statusCode shouldBe HTTP_OK
+            val idpResponse = getServiceProvider(IdpResponder::class)
+                    .getIdpPostResponse(response).decorate()
+            idpResponse.bindingVerifier().verify()
 
-            SingleSignOnProfileVerifier(responseDom).verify()
+            val responseDom = idpResponse.responseDom
+            CoreVerifier(responseDom).verify()
+            ResponseProtocolVerifier(responseDom, TestCommon.ID, acsUrl[SamlProtocol.Binding.HTTP_POST]).verify()
```
Comment (coyotesqrl) [inline test/idp/src/main/kotlin/org/codice/compliance/web/sso/PostLoginTest.kt]: ✏️ More places where static imports would shorten.

File: test/idp/src/main/kotlin/org/codice/compliance/web/sso/RedirectLoginTest.kt [hunk 12]
```diff
@@ -34,27 +33,32 @@ import org.codice.compliance.verification.profile.SingleSignOnProfileVerifier
...
-            return SimpleSign().signUriString(SAML_REQUEST, encodedRequest, relayState)
-        }
-    }
-
     init {
         RestAssured.useRelaxedHTTPSValidation()
 
         "Redirect AuthnRequest Test" {
-            val queryParams = setupAuthnRequest(null)
+            val authnRequest = AuthnRequestBuilder().buildObject().apply {
```
Comment (coyotesqrl) [inline test/idp/src/main/kotlin/org/codice/compliance/web/sso/RedirectLoginTest.kt]: ✏️ Just like in the post tests, these blocks to build the `authnRequest` look suspiciously similar.

File: test/idp/src/main/kotlin/org/codice/compliance/web/sso/RedirectLoginTest.kt [hunk 13]
```diff
@@ -70,20 +74,31 @@ class RedirectLoginTest : StringSpec() {
...
             idpResponse.bindingVerifier().verify()
 
             val responseDom = idpResponse.responseDom
-
             CoreVerifier(responseDom).verify()
-
-            ResponseProtocolVerifier(responseDom, ID).verify()
-
-            SingleSignOnProfileVerifier(responseDom).verify()
+            ResponseProtocolVerifier(responseDom, ID, acsUrl[SamlProtocol.Binding.HTTP_REDIRECT]).verify()
```
Comment (coyotesqrl) [inline test/idp/src/main/kotlin/org/codice/compliance/web/sso/RedirectLoginTest.kt]: ✏️ More places where static imports would shorten.

File: test/idp/src/main/kotlin/org/codice/compliance/web/sso/RedirectLoginTest.kt [hunk 14]
```diff
@@ -107,12 +122,51 @@ class RedirectLoginTest : StringSpec() {
             bindingVerifier.verify()
 
             val responseDom = idpResponse.responseDom
-
             CoreVerifier(responseDom).verify()
+            ResponseProtocolVerifier(responseDom, TestCommon.ID, acsUrl[SamlProtocol.Binding.HTTP_REDIRECT]).verify()
```
Comment (coyotesqrl) [inline test/idp/src/main/kotlin/org/codice/compliance/web/sso/RedirectLoginTest.kt]: ✏️ More places where static imports would shorten.

File: test/common/src/main/kotlin/org/codice/compliance/verification/core/ResponseProtocolVerifier.kt [hunk 15]
```diff
@@ -14,13 +14,18 @@
...
+import org.codice.compliance.SAMLSpecRefMessage.SAMLCore_3_2_2_d
+import org.codice.compliance.SAMLSpecRefMessage.SAMLCore_3_2_2_e
+import org.codice.compliance.SAMLSpecRefMessage.SAMLCore_3_4
 import org.codice.compliance.allChildren
 import org.codice.compliance.children
-import org.codice.compliance.utils.TestCommon.Companion.ACS_URL
 import org.w3c.dom.Node
 
-class ResponseProtocolVerifier(val response: Node, val id: String) {
+class ResponseProtocolVerifier(private val response: Node, val id: String, private val acsUrl: String?) {
```
Comment (bakejeyner) [inline test/common/src/main/kotlin/org/codice/compliance/verification/core/ResponseProtocolVerifier.kt]: ✏️ Might be a good idea to throw the extra values here into the decorators.

File: test/idp/src/main/kotlin/org/codice/compliance/web/sso/PostLoginTest.kt [hunk 16]
```diff
@@ -91,10 +118,47 @@ class PostLoginTest : StringSpec() {
...
-            ResponseProtocolVerifier(responseDom, TestCommon.ID).verify()
+            response.statusCode shouldBe HTTP_OK
+            val idpResponse = getServiceProvider(IdpResponder::class)
+                    .getIdpPostResponse(response).decorate()
+            idpResponse.bindingVerifier().verify()
 
-            SingleSignOnProfileVerifier(responseDom).verify()
+            val responseDom = idpResponse.responseDom
+            CoreVerifier(responseDom).verify()
+            ResponseProtocolVerifier(responseDom, TestCommon.ID, acsUrl[HTTP_POST]).verify()
```
Comment (bakejeyner) [inline test/idp/src/main/kotlin/org/codice/compliance/web/sso/PostLoginTest.kt]: ✏️ See comment above about decorators.

**Pull request merged**